drivers: i2c_sam0: Optimize and add i2c slave

sslupsky opened the pull request:
Optimizations:
When using system power management, the bus transactions
can be interrupted by low power modes which causes excessive
power consumption and can cause the transactions to terminate
before completed. Further, other interrupts can cause bus transaction
anomalies.

When using the RESTART flag, the MB and SB bits must be cleared.

This commit prevents interruption of the bus during low power modes and
correctly clears the MB or SB flags when a bus transaction is restarted.
Also fixes an initialization problem with the INACTOUT timer and ensures
the LOWTOUT is properly configured when using i2c_configure().

Add i2c slave support:
The existing sam0 i2c driver does not support slave mode.

This commit enables slave support for sam0 devices.

The Slave support is based on Linux I2C Slave support by Wolfram Sang
and documented at
https://www.kernel.org/doc/Documentation/i2c/slave-interface

The code for this commit is largely based on code taken from PR #5224
and the stm mcu implementation. This was modified to work with the
sam0 SERCOM perhiperal and integrated into the sam0 driver.

Fix whitespace and comments

This PR replaces https://github.com/zephyrproject-rtos/zephyr/pull/24047

Fixes #21711
Fixes #21549
Fixes #21233
Fixes #21232
Fixes #21114
Fixes #21092

sslupsky committed:
drivers: i2c_sam0: Optimize i2c

Optimizations:
When using system power management, the bus transactions
can be interrupted by low power modes which causes excessive
power consumption and can cause the transactions to terminate
before completed. Further, other interrupts can cause bus transaction
anomalies.

When using the RESTART flag, the MB and SB bits must be cleared.

This commit prevents interruption of the bus during low power modes and
correctly clears the MB or SB flags when a bus transaction is restarted.
Also fixes an initialization problem with the INACTOUT timer and ensures
the LOWTOUT is properly configured when using i2c_configure().

The GCLK_SERCOM_SLOW clock is used to time the i2c
time-outs and must be configured to use a 32KHz oscillator.
See 28.6.3.1

This commit configures GCLK4 to 32768 Hz

Fixes #21711, #21549, #21233, #21232, #21114, #21092

A problem can arise when back to back transactions occur before
the STOP transaction is complete.

This commit also introduces a busy wait for the STOP to complete
with an timeout in the event there is a bus failure.

Signed-off-by: Steven Slupsky <sslupsky@gmail.com>

sslupsky commented:
There is an issue when two transactions run back to back.  I added an idle loop and timeout to address this.

pabigot commented:
@sslupsky Could you revisit this PR, please?

First the single commit has far too much stuff in it.  It should be split out into separate commits that fix each isolated problem separately.  (It looks like it was that way originally, but then got squashed; I believe that was a misunderstanding of expectations that's been corrected.)

We're particularly interested in #21232 which is medium priority so affects the release.  Ideally a fix for that would be pulled out to a separate PR that we can merge: this one has too much other stuff in it.

I don't see how this PR fixes #21114; if it doesn't please remove it from the list of fixed issues.  (I edited the description so that all of them, not just the first on the line, were recognized by github as related.)

sslupsky commented:
@pabigot ok, i can split this up.

Regarding #21114, i believe the erratic timeouts were because the i2c driver was doing some very weird things.  Since fixing this driver and the rtc, I no longer see the erratic timeouts.

carlescufi commented:
@sslupsky I would be most grateful if you could split the fix to  #21232 in a separate PR as soon as possible. Thanks!

nandojve commented:
@sslupsky could you rebase.

sslupsky commented:
@nandojve Against what?  master?  I haven't updated my local to the most recent master since release of 2.3 yet.

nandojve commented:
> @nandojve Against what? master? I haven't updated my local to the most recent master since release of 2.3 yet.

Well, there is a file conflict, at least for me shows below: 
I hope you can update soon.
![image](https://user-images.githubusercontent.com/7027887/86154416-6d248180-bad9-11ea-9bf2-133daf2a2792.png)

sslupsky commented:
Ok, I will have a look.  I am filtering out the commit for the i2c slave changes and will push the updated PR shortly.

sslupsky committed:
drivers: i2c_sam0.c: add i2c slave

The existing sam0 i2c driver does not support slave mode.

This commit enables slave support for sam0 devices.

The Slave support is based on Linux I2C Slave support by Wolfram Sang
and documented at
https://www.kernel.org/doc/Documentation/i2c/slave-interface

The code for this commit is largely based on code taken from PR #5224
and the stm mcu implementation. This was modified to work with the
sam0 SERCOM perhiperal and integrated into the sam0 driver.

A slave must wait for data ready when there is an address match
to ensure the data is valid.

Disable SCLSM (SCL stetch mode) when operating in slave mode
to ensure that i2c ACK's are handled properly.

Fix k_timeout_t.

Signed-off-by: Steven Slupsky <sslupsky@gmail.com>

sslupsky commented:
@nandojve @pabigot I pulled the i2c slave changes into a separate commit.

sslupsky commented:
I added a  commit for some changes made since the original PR was submitted.

sslupsky commented:
hmm, I forced pushed the changes but it seems to have only updated the commit dealing with the i2c_slave.  Can someone please advise how to push the changes to the earlier commit (I squashed those changes).

sslupsky commented:
Ok.  I think I have properly rebased this now and the two commits accurately separate the i2c slave changes.

dleach02 commented:
needs a review from the relevant folks

nandojve commented:
Hi @sslupsky, you are doing a great job, thanks!

I would like run this on HW and I have two R21 and one D20 that I can use it. How do you suggests to do it? How do you are testing, loopback? Do you have anything that can share to speed up my tests?

I would like know what app is building this on CI, do you know one? We need make sure code is covered on CI.

pabigot reviewed (changes requested):
This PR needs to be updated for the change to replace `u32_t` with `uint32_t` back in June.  A pinned message in the general slack channel has these instructions, which would need to be adapted now.

```
git checkout ${BR}
git rebase e77985a2f0 # last commit before rename
rm -rf c99
git format-patch -o c99 e77985a2f0..HEAD 
sed -i.bak -r \
  -e 's/\bu(8|16|32|64)_t\b/uint\1_t/g' \
  -e 's/\bs(8|16|32|64)_t\b/int\1_t/g' \
  c99/*
for f in c99/*.patch ; do
  diff -u ${f}.bak ${f}
done
git branch BU/${BR}
git reset --hard a2d4292a96  # the last commit of the rename
git am c99/*.patch
```

sslupsky commented:
@nandojve I have a couple boards that I have been using for my testing.  Mostly, I have tested with a few sensors on my board  (temp/humidity, barometer, ambient light, accelerometer).  I have another board to board connection where two D21's talk to each other using I2C as well.  The sensor application periodically reads the sensors and the application on the other board pulls the sensor readings from the sensor board using the AT24 memory device emulation driver.

I do not have a dedicated test suite so there is no CI set up.
